lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -856,199 +870,238 @@ void mlir::triton::updateWaits(ModuleOp module) {
+            operand.getType().cast<TensorOrMemDesc>().getEncoding())) {
+      Value transitiveOperand = operand;
+      while (isa_and_nonnull<ttg::ConvertLayoutOp, tt::TransOp>(
+          transitiveOperand.getDefiningOp())) {
+        transitiveOperand = transitiveOperand.getDefiningOp()->getOperand(0);
+      }
+      return forOp.isDefinedOutsideOfLoop(transitiveOperand) ||
+             isa<ttg::MemDescSubviewOp>(transitiveOperand.getDefiningOp());
+    }
+
+    return true;
+  };
+
+  if (!checkOperand(dotOp.getA()) || !checkOperand(dotOp.getB()) ||
+      !checkOperand(dotOp.getC()))
+    return std::nullopt;
+
+  // Rule 2: The dot should only be used by the for loop's `yield`.
+  if (!dotOp->hasOneUse() || !isa<scf::YieldOp>(*dotOp->getUsers().begin()))
+    return std::nullopt;
+
+  // The result of the dot becomes this loop carry value.
+  auto iterArgIdx = dotOp->getUses().begin()->getOperandNumber();
+
+  // Rule 3: The only users of the dot's result from iteration i-1 are other
+  // MMAv3 dots (synchronous or async) as the `c` operand.
+  if (!llvm::all_of(forOp.getRegionIterArg(iterArgIdx).getUses(),
+                    [&](OpOperand &use) {
+                      return dynCastMMAv3Dot(use.getOwner()) &&
+                             use.getOperandNumber() == 2;
+                    })) {
+    return std::nullopt;
   }
+
+  return iterArgIdx;
 }
 
-void mlir::triton::asyncLaunchDots(scf::ForOp forOp) {
-  Block *loop = forOp.getBody();
-  auto getBlockNumInFor = [](Operation *op, scf::ForOp forOp) {
-    if (!op)
-      return -1l;
-    auto lastOp = op;
-    while (op->getBlock()->getParentOp() != forOp) {
-      lastOp = op;
-      op = op->getBlock()->getParentOp();
+// If necessary, insert a dot-wait inside the loop, waiting for the results of
+// the async dots from iteration i-2 to complete.  (We pipeline to depth 2, so
+// there are at most 2 dots in flight at any point in time.)
+//
+// TODO(jlebar): I'm told CUTLASS pipelines to depth 4 or even 8, to better hide
+// variable-latency operations.  Would this help us?
+//
+// We can skip inserting the wait if we have a synchronous MMAv3 dot that
+// appears either (a) before any uses of async dots' values from iteration i-1,
+// or (b) after all uses of the async dots in iteration this iteration.
+//
+// By contract, the only users of the async dots' results from iteration i-1 are
+// other MMAv3 dots, and the only user of the async dots in this iteration is
+// the loop's `yield` op.  So it's sufficient to check whether a sync dot
+// appears first or last in the list of all sync+async dots.
+//
+// If such a sync dot exists, we convert it to async+wait so that we can thread
+// the dependencies through the wait op.  This keeps other passes from
+// reordering the dots.
+//
+// TODO(jlebar): It's possible that the sync dot might appear in the wrong place
+// but could be hoisted/sunk to the right place.  Right now we don't handle
+// this.
```

Comment:
I was thinking of something like this.

%a = dot.async %x
%b = dot %y
%c = dot.async %b

The second dot has to be synchronous because it's used by the third dot.  But the first two dots do not depend on each other, so we can do them in any order.  It would therefore be better to do %b first and then do %a -- that way we could run %a and %c at the same time.

Does that not work?

Reply:
I don't see how you'd do it beforehand, because until this pass runs, you don't know which dots are going to be sync vs async?  And if you did it afterwards, you'd have to repeat all of the work this pass does to determine if it's safe to move?